agent/eni/watcher_linux.go

```diff
@@ -0,0 +1,194 @@
+	}
+
+	// Pass state to Init
+	udevWatcher.state.Init(links)
+	return nil
+}
+
+// Start periodically updates the state of ENIs connected to the system
+func (udevWatcher *UdevWatcher) Start() {
+	// UDev Event Handler
+	go udevWatcher.eventHandler(udevWatcher.ctx)
+	udevWatcher.performPeriodicReconciliation(udevWatcher.ctx, defaultReconciliationInterval)
+}
+
+// Stop is used to invoke the cancellation routine
+func (udevWatcher *UdevWatcher) Stop() {
+	udevWatcher.cancel()
+}
+
+// performPeriodicReconciliation is used to periodically invoke the
+// reconciliation process based on a ticker
+func (udevWatcher *UdevWatcher) performPeriodicReconciliation(ctx context.Context, updateInterval time.Duration) {
+	udevWatcher.updateIntervalTicker = time.NewTicker(updateInterval)
+	for {
+		select {
+		case <-udevWatcher.updateIntervalTicker.C:
+			udevWatcher.reconcileOnce()
+		case <-ctx.Done():
+			udevWatcher.updateIntervalTicker.Stop()
+			return
+		}
+	}
+}
+
+// reconcileOnce is used to reconcile the state of ENIs attached to the instance
+func (udevWatcher *UdevWatcher) reconcileOnce() {
+	log.Debugf("Udev watcher reconciliation: begin")
+	links, err := udevWatcher.netlinkClient.LinkList()
+	if err != nil {
+		log.Warnf("Udev watcher reconciliation: error retrieving network interfaces: %v", err)
+		return
+	}
+
+	// Return on empty list
+	if len(links) == 0 {
+		log.Info("Udev watcher reconciliation: no network interfaces discovered for reconciliation")
+		return
+	}
+
+	currentState := udevWatcher.buildState(links)
+
+	// NOTE: For correct semantics, this entire function needs to be locked.
+	// As we postulate the netlinkClient.LinkList() call to be expensive, we allow
+	// the race here. The state would be corrected during the next reconciliation loop.
+
+	udevWatcher.state.Reconcile(currentState)
+	log.Debugf("Udev watcher reconciliation: end")
+}
+
+// IsMACAddressPresent checks if the MACAddress belongs to the maintained state
+func (udevWatcher *UdevWatcher) IsMACAddressPresent(macAddress string) bool {
+	return udevWatcher.state.IsMACAddressPresent(macAddress)
+}
+
+// getAllENIs is used to retrieve the state observed by the Watcher
+func (udevWatcher *UdevWatcher) getAllENIs() map[string]string {
+	return udevWatcher.state.GetAll()
+}
+
+// buildState is used to build a state of the system for reconciliation
+func (udevWatcher *UdevWatcher) buildState(links []netlink.Link) map[string]string {
+	state := make(map[string]string, mapCapacityHint)
+
+	for _, link := range links {
+		macAddress := link.Attrs().HardwareAddr.String()
+		if macAddress != "" {
+			state[macAddress] = link.Attrs().Name
+		}
+	}
+	return state
+}
+
+// eventHandler is used to manage udev net subsystem events to add/remove interfaces
+func (udevWatcher *UdevWatcher) eventHandler(ctx context.Context) {
+	udevWatcher.udevMonitor.Monitor(udevWatcher.events)
+	for {
+		select {
+		case event := <-udevWatcher.events:
+			subsystem, ok := event.Env[udevSubsystem]
```

Comment:
Since there are some events doesn't fall into add|remove, it's good to add an debug level log here, which may provide helpful information for debugging in the future, especially when something changed in the event format.